docs/src/proposals/interest-bearing-tokens.md

```diff
@@ -0,0 +1,284 @@
+
+Programs and wallets that wish to support i-tokens must update the instructions
+to use all `Checked` variants and use the new read-only instructions to fetch
+information.
+
+The SPL token library provides wrappers compatible with any conforming program.
+For the base SPL token program, the wrapper simply deserializes the account data
+and returns the relevant data. For other programs, it calls the appropriate
+read-only instruction using an ephemeral account input.
+
+The structure for new holding and mint accounts must follow the layout of the
+existing SPL token accounts to ensure compatibility in RPC when fetching pre and
+post token balances.
+
+For a holding, the first 165 bytes must contain the same information
+as a normal SPL token holding. The following byte must be the type of
+account (ie. `Holding`), and after that, any data is allowed as required by the
+new token program.
+
+The same applies for mints, but only for the first 82 bytes.
+
+The `amount` field in an SPL token holding should contain a balance in order
+to properly integrate with RPC's `preTokenBalances` and `postTokenBalances`.
+
+### Token Program Registry
+
+We need the [token-list](https://github.com/solana-labs/token-list) to include
+vetted SPL token-conforming programs.
+
+TODO consider adding Rust and C versions of the registry for on-chain programs and RPC
+
+TODO consider detailing the process of including a new program
+
+### Runtime
+
+#### Ephemeral Accounts
+
+In order to reduce friction when using read-only instructions on other programs,
+we need the ability for on-chain programs to dynamically create "ephemeral
+accounts", which only contain data, can be passed through CPI to other programs,
+and disappear after the instruction.
+
+For a program using i-tokens, to get the balance of a holding, the program creates
+an ephemeral account with 8 bytes of data, passes it to the i-token program's
+`GetHoldingBalance` instruction, then reads the information back as a `u64`. At the
+end of the instruction, the account disappears.
+
+TODO What's the best approach to implement this?  For example, does an ephemeral
+account even need a public key?
+
+This solution avoids the need to pass additional scratch accounts to almost all
+programs that use tokens.
+
+TODO what limits do we consider for these? A certain number of accounts?
+A total amount of bytes? Regarding security, this breaks the previous
+check that a CPI must use a subset of accounts provided to the calling program.
+
+#### Dynamic sysvars
+
+The i-token program also requires the use of dynamic sysvars. For example,
+during the `GetHoldingBalance` instruction, the program needs to know the current
+time in order to properly convert from token amount to share amount.
+
+Without dynamic sysvars, we need to create a system to tell what sysvars are
+needed for different instructions on each program, which creates complications
+for all users.
```

Comment:
This is my very wild idea but, I'm thinking it's about time for us to realize some dynamic programmable `AccountInfo`s creation for composable TXes:

_use preflight transaction execution for dynamic AccountInfo generation_

So, when tx is executed by preflight in the rpc, it records actually dynamically referenced accounts and returns the list to the client. Then, client resends the real tx to the rpc with added AccountInfos. 

```rust
fn entrypoint(account_infos: DynamicAccountInfos) {
  account_info.get(pubkey, is_readonly, is_signer)?
  let inner_account_info = account_info.create_inner_account_info();
  inner_account_info.push(...);
  ...
  invoke(ix, &inner_account_info);
}

impl DynamicAccountInfos {
  // in pre-flight simulation execution environment, record requested account keys into HashMap and modes then actually return the account data via syscall
  // in real execution environment, return the account if tx referenced it correctly; otherwise return Err
  fn get(...) -> Result<AccountInfo> {
  }
  // create AccountInfo with empty positional accounts, while cloning the HashMap for the dynamically-referenced accounts; this is like ARGV/Env relation.
  fn create_inner_account_info()
  }
}

```


pro: downstream developer of any defi app doesn't need care about which accounts to pass, minimum friction for writing newer code and maximum compatibility is retained, because pre-flight is done off-chain, relatively less load on the on-chain processing. Also, this enables vectorization of write-referenced accounts. Lastly, greatly flexibility at the hands of program devs
con: 1 extra round trip between client and rpc. Offline tx is a bit more tedious. This can be abused?

Or if we allow delaying and buffering the tx internally inside bank (step 1: pre-flight, step 2: actual execution):

pro: no round-trip at the client side.
con: harder proper cost structure, 2x load on banking/replaying path (also optimistic locking can be employed to minimize the load?)

Reply:
Putting aside customized ntfs, my motivating use case was erc20's paxg, which collects its special small fees everytime its token is transferred.

Reply:
This is quite an idea!  We definitely need some model for declaring the required accounts in an instruction.  This is a cool solution for accounts where the pubkey is known.  That covers sysvars, programs, global state accounts, dependent accounts (e.g. the mint associated with a token swap), but would not cover certain positional arguments, correct? How would this work with positional arguments where we don't know the pubkey?

Reply:
> would not cover certain positional arguments, correct? How would this work with positional arguments where we don't know the pubkey?

Seems my idea is grokked by you, correctly. this is a good question. :) Well, this is possible. thanks to the force of blockchain's deterministic behavior.
The reason is 2 phased:

Firstly, it's always determinable to dynamically know the actual needed pubkeys at the time of simulation from runtime perspective. That's because conceptually, tx must have been constructed with priori knowledge of correct understanding for positional parameters.

So, consider like this (maybe you had this code in your mind?):

```
     let foo_bar = FooBar::unpack(account_info.next());
    foo_bar.do_baz();
```

Given this code, the right pubkey for the positional account can't be known by runtime simulation. But this is fine. The caller (another cpi-ed program or ultimate defi web browser app) must know how to explicitly specify the right pubkey. Any semantically wrong pubkey will lead to `Err`. Conversely, if the program did work by passing arbitrary account at the position, then the account is extra (unneeded) positional argument or just the program is badly written.

Secondary, let's think about this from theoretical stand point. There's two key roles here. program author and tx creator. Ultimately, both parties just want to ensure expected behavior for actual state transitions in the form of `s' = program(s, tx) where s = list(account)`. program author want program to refer to accounts for its internal book-keeping but also parameterize some ones as input of the state transition _to be a actually useful program to be used by other parties_. In other words, the parametarized accounts are unknownable up to the actual execution. So, at this moment, list(account) could contain unknown elements. Then, tx creator alike want the program to refer to accounts for their utility using their knowledge of expected behavior of the program (Also don't want to know which internal account is needed for the program's book-keeping!). At this moment, the list(account) should be sufficient set of accounts to realize the deterministic expected behavior, containing no undetermined account. Otherwise, program author or tx creator did wrong.

In other words, we can categorize AccountInfos into two kinds: static account info, dynamic account info:
static account info: determined by program author; ultimately, all can be determined by runtime simulation.
dynamic account info: determined by tx creator with their knowledge.

Also, the good thing for the dynamic account list generation is that, it's not actually security concern to allow the program author to add arbitrary number of account info to the tx thanks to our signer bit and strict account ownership runtime mechanism. After all, all program can read ( but not write) any state of accounts as before. :)

Reply:
I think this is out of scope for this proposal, which is already quite large and highly impactful on the ecosystem.  Overhauling how accounts work should be separate and if done before i-tokens could be used here as well.

Reply:
I see what you mean here.  I was worried more specifically about dynamic *nested* accounts.  For example, in the case of token lending, you need the lending reserve as a dynamic input, but inside of that are all of the extra accounts *inside* the reserve, like the token mint for that reserve.  We can say that a tx creator should just know all of that up front, but since we're talking about a model that declares everything that you'll need up front, it would be extremely useful to fetch dependent accounts as well.  I think your model could possibly be extended to make this work though :-)

Either way, since this is out of scope for the current proposal, I will resolve the comment.  Thanks a bunch!

Reply:
> I was worried more specifically about dynamic nested accounts. For example, in the case of token lending, you need the lending reserve as a dynamic input, but inside of that are all of the extra accounts inside the reserve, like the token mint for that reserve. ... I think your model could possibly be extended to make this work though

@joncinque I think the current model actually works for dynamic nested accounts because the tx simulation also scans any nested CPI for on-demand account scanning and those accounts are implicitly passed to the nested InvokeContext like environment variables. This being said, existing programs are need to be modified to tell the address hint to the runtime at least. Also, I'll try to slip this facility in the timeline of itoken transition across the board.

Thanks to #16218, this will be kind of easy, I hope.

Reply:
@ryoqun this all depends on dynamic account loading in the simulation mode, correct?  It looks like we do have a model tht works then :smile:

Reply:
wow, 2 months has passed.. it've got some time to write this finally: https://github.com/solana-labs/solana/pull/17796